src/librustc_typeck/check/mod.rs

```diff
@@ -1306,6 +1306,24 @@ fn check_union<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     check_packed(tcx, span, def_id);
 }
 
+fn check_opaque<'a, 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    def_id: DefId,
+    substs: &'tcx Substs<'tcx>,
+    span: Span,
+) {
+    if let Err(partially_expanded_type) = tcx.try_expand_impl_trait_type(def_id, substs) {
+        let mut err = tcx.sess.struct_span_err(span, "recursive opaque type");
```

Comment:
Maybe this deserves a dedicated E0XXX error code? (which would be defined in librustc_typeck/diagnostics.rs)